### DISCUSSION TITLE
Real Time data acquisition preventing menu tasks from running

### DISCUSSION THREAD

**Discussion by leelx78** [davetcc/tcMenu#413] [General]
Hi Dave,

I'm working on a project where my ESP32-S3 CPU is maxed out doing real-time data processing in the loop() function on Core 1, together with the Arduino system. This means:
```
void setup() {
    DEVICE.begin();
}

void loop() {
    DEVICE.loopFunction();
}
```
...where DEVICE is my class with a begin() and loopFunction() methods.

Then I wanted to add TCMenu and a SSD1306 display plus a rotary encoder/switch but it doesn't seem to work at all. the function `taskManager.runLoop();` seems blocking or just uses too much CPU time. I tried 2 approaches i.e.

```
void setup() {
    setupMenu();
    DEVICE.begin();
}

void loop() {
    DEVICE.loopFunction();
    taskManager.runLoop();
}
```
and
```
void setup() {
    setupMenu();
    DEVICE.begin();
    taskManager.schedule(repeatMicros(100), [] {
      DEVICE.loopFunction();
    });
}

void loop() {
    taskManager.runLoop();
}
```

With the same result.

Do you have any suggestions? In your video I remember you mentioned running a second  taskManager on another core but I didn't see how to do it. Ideally I'd like to keep the UI on core 0 and all the existing real-time processing on core 1

Let me know what you think

thanks!

**Comment by davetcc** [comment]
Please only raise actual issues using issues   Anything else create a discussion. The format of issues must not be changed and they are always to be created using templates. Thanks.

**Comment by davetcc** [comment]
First I would confirm that everything works properly without the resource consuming task in play. IE in a test project. 

Then as you say you could try to move that resource consuming task to another core. Using TaskManagerIO as a queue for example. That should be as simple as submitting work to it (maybe using the  execute method). Be aware that any data shared between cores will need to be synchronized or atomically safe.   Steps I can think of are:

* globally create a second task manager instance. 
* create another native task on the other core
* Run the task manager loop within that. 
* Submit work to be done on the other taskManager.

**Reply by leelx78** [reply]
Hi Dave I'm still having loads of issues in moving anything to the other core (so many crashes!).

Is there any way to bypass the taskManager and call the screen refresh manually? I have no problem is displaying logos and changing text on SSD1306 via the "SSD1306Wire" library while data acquisition is running and on the same core. I just call the oled refresh function every 200ms or so and it works fine.

Of course there is no menu that way and it looks a lot more basic but this was just as a proof of principle.

Any other ideas? Anything else that can slow things down?

**Reply by davetcc** [reply]
> still having loads of issues in moving anything to the other core (so many crashes!).

Yep, that is a highly advanced topic and you need to be certain what methods are called on the other core. 

> Anything else that can slow things down?

what’s the actual issue. IE if you remove all heavy operations does it work. How does it fail when the device is added? How long does the device loop take to execute? Is it long enough to be making switches miss the rotary encoder turns.

**Reply by davetcc** [reply]
Don’t forget taskManager is an unfair scheduler. What ever comes first comes first, and  something that always jumps in at the front of the queue then it wins. You need to make sure that there is chance for others to run.

**Reply by davetcc** [reply]
I think there are a few options here that would not affect task manager, if something genuinely has a need to run in such a tight loop, one option is to add it to the `loop()` method, IE something like:

```
// rest of sketch removed..
void loop() {
   taskManager.runLoop();
   DEVICE.runLoop();
}
```

But there are some things to consider, the display you're using is not exactly the fastest display, it is probably an I2C display, and on Arduino I2C operations are blocking, meaning that your sketch will be frozen until the data is transferred. One way to reduce that may be to use an SPI display instead and crank the rate up as far as the display will allow it. 

For example, a common issue is that people have issues when trying to drive steppers at the same time as other operations, for me, it would be better to write some native interrupt/timer code that handled the stepper separately, but I say that as someone very used to writing mutlithreaded/synchronized/atomic code. There are no easy answers here, but in your case, I'd maybe try the above first and see how that works.

**Reply by leelx78** [reply]
Thanks @davetcc the thing is that with the SSD1306Wire library it seems to work fine (e.g. updating text or drawing a logo every 100-200ms or so) so I doubt there is a I2C blocking issue unless the library you use has a ton of consecutive I/O that way. I have a buffer in the data acquisition routine so if I have infrequent operations taking a bit of time it still works fine overall.

Is there a way to explicitly calling the screen refresh at given intervals and minimising any work in between? 

Anything else that might be blocking in the rest of the code?

**Reply by davetcc** [reply]
>  I doubt there is a I2C blocking issue unless the library you use has a ton of consecutive I/O that way

Unfortunately, on nearly all platforms even some native I2C implementations are blocking, and for OLEDs the protocol is chatty so yes it will block. I normally use SPI screens to avoid this but for OLED the protocol is still chatty.

As per comment from last week, TaskManagerIIO is an unfair scheduler, it will schedule the same task/event over and over if it is at the front of the queue. If you schedule something every 100 micros that takes more than 100 micros to run, it may end up taking over the queue and always running.  This is a common problem in co-routine based systems, and one way around it would be to update your device in the runloop after the taskManager.runLoop call. Have you tried that as I presented above?

**Reply by davetcc** [reply]
So I think at this point we need clarity on what the problem is:

1. What the exact outstanding problem is.
2. That (if appropriate to that problem) you've  tried the above suggestion of moving the device run loop to the loop method.

**Reply by leelx78** [reply]
thanks @davetcc I have another project this week but I'll get in touch as soon as I have a chance to test this further

**Reply by leelx78** [reply]
Hi @davetcc I think I narrowed down the issue to the rotary interface, it seems it doesn't like the way it polls or uses interrupts as it disturbs the rest of the code somehow.

I implemented my own interface to the encoder which works well so far and correctly creates events like
```
void onClockwiseRotation(){
}
void onCounterClockwiseRotation(){
}
void onEncoderPressed(){
}

```
and I had to select "No Input Required" in TC Menu Designer. 

Now, what's the best way to pass these events to the menu so I can navigate it?

**Reply by leelx78** [reply]
I tried calling these in the events but don't seem to update anything
```
menuMgr.performDirectionMove(false);
menuMgr.performDirectionMove(true);
menuMgr.onMenuSelect(true);
```

**Reply by leelx78** [reply]
works now, I had to initialise the encoder with unused pins and then call manually. Seems like a hack but works.

I'm curious why my encoder interface works so much better though. it's based on the CFRotaryEncoder library, maybe it's worth you have a quick look to see if it's something that might be an option for the main code?

**Reply by leelx78** [reply]
..actually I see that menuMgr.performDirectionMove() is not doing what I need, which is up/down.

@davetcc did I miss something in the MenuManager docs? `menuMgr.onMenuSelect` works well when connected to my encoder interface and `menuMgr.performDirectionMove` also "works" but it only as a back/forward, not up/down.

Maybe I'm overcomplicating things but I tried so many combinations of things, including the `TC_LEGACY_ENCODER` flag and it seems to miss loads of encoder clicks with your interface but it's spot-on with mine for some reasons..

**Reply by davetcc** [reply]
For TC_LEGACY_ENCODER you must be using the unreleased master branch? This is as yet unfinished, and I have yet to test it with Interrupts enabled - early indications show a bug with interrupt support that will need to be worked through before I can release it. 

However, if you really want to handle the encoder yourself, the best way is to write an implementation of the RotaryEncoder class. You could take a look at the `HardwareRotaryEncoder` and `EncoderUpDownButtons` classes and create an implementation for your case.

**Reply by davetcc** [reply]
EG - https://github.com/davetcc/IoAbstraction/blob/master/src/SwitchInput.h#L203

**Reply by leelx78** [reply]
Thanks @davetcc I'll try to implement it as soon as I have a chance.  

A way to do this via MenuManager would be handy for the future if you have a chance though. Something like `menuMgr.performDirectionMoveUp`, `menuMgr.performDirectionMoveDown` and so forth.

**Comment by leelx78** [comment]
Hi @davetcc I close this and open a new issue since I'm narrowing it down to the encoder class thanks
